### PR TITLE
Add string tensor tags to new summary interface

### PR DESCRIPTION
**Pull request by thuyen**
Old/Deprecated summary ops can accept `string tensors` for `tags`. The new summary interface only allows `strings` for `names` (subsequently used as tags). This pull adds `prefix` option, which can be a `string tensor`. 

This is important because sometimes we want to define the tags symbolically based on some conditions of input placeholders.

### CODE REVIEW

**Comment by thuyen** [comment]
@danmane @drpngx Can one of the admins have a look at this pull request?

**Comment by drpngx** [comment]
@danmane knows best.

**Comment by danmane** [comment]
Is there a need for the "prefix" argument? Why not just have the summary wrapper accept a string tensor as name, and convert it into a string?

Adding prefix as another arg seems orthogonal to the main goal of the pr

**Comment by thuyen** [comment]
@danmane Could you tell me the function for `string_tensor to string` conversion. I couldn't find it in current API. If we have that function then we don't need this PR at all.

**Comment by thuyen** [comment]
I am still not sure if such function exists since `tensor` value would be determined dynamically at run time while `string` needs static literals.

**Comment by martinwicke** [comment]
What would the prefix be used for? Can't you achieve the same by wrapping the call to the summary in a name_scope?

**Comment by thuyen** [comment]
`name_scope` couldn't be a `string tensor`. 
With the old interface, one can create a single graph for both training and validation (like the following). In the new interface, it's only possible with some new argument `prefix` (a `tensor`) because the `name` argument can only be `string`. An example:
```
is_training = tf.placeholder(tf.bool)
inputs, labels = data(is_training)
outputs = model(inputs, is_training)
loss = tf.reduce_mean(tf.square(output-labels))
train_op = ....

prefix = tf.cond(is_training, lambda: tf.constant('train/'), lambda: tf.constant('valid/'))
tf.summary.scalar('loss', loss, prefix=prefix)
summary_op = tf.summary.merge_all()

# Each training iteration
sess.run([train_op, summary_op], feed_dict={is_training: True})

# Validation
sess.run(summary_op, feed_dict={is_training: False})
```

**Comment by foxik** [comment]
+1 for implementing some way of using string tensor as summary name.

I frequently used the following scenario, where I want to measure accuracy on train/devel/test data (to appear in different graphs in TensorBoard):
```python
    dataset_name = tf.placeholder(tf.string, [])
    summary = tf.scalar_summary(self.dataset_name+"/accuracy", accuracy)
```
That way I could use the same `summary` with any of the data:
```python
   summary = session.run([summary], {dataset_name:"train" / "devel" / "test"})
```
A `prefix` option would be ideal for me, but generaly any way of using string tensor is. Currently I have to create separate summary operations for every dataset I want to evaluate on.

**Comment by ppwwyyxx** [comment]
+1 for passing an optional string/string tensor (but not prefix) as names. This would also resolve #6150

**Comment by drpngx** [comment]
@dandelionmane is that the right way to go about this?

**Comment by rmlarsen** [comment]
@dandelionmane any advice on how to proceed here?

**Comment by dandelionmane** [comment]
@thuyen Thanks for the pull request, and sorry for the really slow review on my part.

I think this is a fine proposal. I would just like more descriptive docstrings that explain why the new argument should be used.

File: tensorflow/python/summary/summary.py [hunk 1]
```diff
@@ -104,6 +104,7 @@ def scalar(name, tensor, collections=None):
     tensor: A real numeric Tensor containing a single value.
     collections: Optional list of graph collections keys. The new summary op is
       added to these collections. Defaults to `[GraphKeys.SUMMARIES]`.
+    prefix: Optional string tensor
```
Comment (dandelionmane) [inline tensorflow/python/summary/summary.py]: Optional string tensor. If provided, it will be a prefix to the summary's tag in TensorBoard.

**Review by teamdandelion — CHANGES_REQUESTED** [review]
I'd like the docstring for "prefix" to explain how the prefix will be used. Otherwise, LGTM

**Comment by vrv** [comment]
Adding @martinwicke for API review triage, if needed.

Also, a test is warranted here, probably.

**Comment by dandelionmane** [comment]
Please update it for the (tensor_summary)[https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/summary_ops.py#L38] as well. This will be more tricky because the tensor_summary has a different implementation and things are a little inconsistent. 
Please also add tests.

In the long run, I think we may want to have all of the summaries set the node_name part of the summary protobuf according to the kernels' node name, and set the tag according to things like the prefix and the original text the user provided to the summary op (as opposed to the deduplicated, name-scope'd name)

**Comment by ppwwyyxx** [comment]
Why the solution here is prefix instead of just giving the whole string? Giving the whole string would also fix #6150 together

**Comment by martinwicke** [comment]
@dandelionmane, what about @ppwwyyxx's comment?

**Comment by drpngx** [comment]
@dandelionmane any progress?

**Comment by thuyen** [comment]
@dandelionmane `tensor_summary` has different interface, taking only the `tensor` as [input](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/logging_ops.cc#L68), no `tag`, and thus couldn't take a string tensor as tag. This is probably inline with the new summary interface . I should wait for more unified solution from google and close this?
